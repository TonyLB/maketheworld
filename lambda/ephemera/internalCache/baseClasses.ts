import { EphemeraCharacterId, EphemeraRoomId, LegalCharacterColor } from "@tonylb/mtw-interfaces/dist/baseClasses";

export class Deferred <T>{
    invalidationCounter: number;
    promise: Promise<T>;
    _resolve: (value: T) => void = () => {}
    _reject: () => void = () => {}
    isFulfilled: boolean = false
    isRejected: boolean = false
    constructor() {
        this.invalidationCounter = 0
        this.promise = new Promise((resolve, reject)=> {
            this._reject = reject
            this._resolve = resolve
        })
    }

    invalidate() {
        this.invalidationCounter += 1
        if (this.isFulfilled || this.isRejected) {
            this.promise = new Promise((resolve, reject)=> {
                this._reject = reject
                this._resolve = resolve
            })
            this.isFulfilled = false
            this.isRejected = false
        }
    }

    resolve (invalidationCounter: number, value: T): boolean {
        if (invalidationCounter < this.invalidationCounter) {
            return false
        }
        this._resolve(value)
        this.isFulfilled = true
        return true
    }

    reject (invalidationCounter: number): void {
        if (invalidationCounter >= this.invalidationCounter) {
            this._reject()
            this.isRejected = true
        }
    }
}

export type RoomCharacterListItem = {
    EphemeraId: EphemeraCharacterId;
    Color?: LegalCharacterColor;
    ConnectionIds: string[];
    fileURL?: string;
    Name: string;
}

export type DependencyEdge = {
    EphemeraId: string;
    key?: string;
    assets: string[];
}

export type DependencyNode = {
    EphemeraId: string;
    completeness: 'Partial' | 'Complete';
    connections: DependencyEdge[]
}

export type LegalDependencyTag = 'Asset' | 'Variable' | 'Computed' | 'Room' | 'Feature' | 'Map'
export const isLegalDependencyTag = (tag: string): tag is LegalDependencyTag => (['Asset', 'Variable', 'Computed', 'Room', 'Feature', 'Map'].includes(tag))

export type DependencyGraphPut = {
    EphemeraId: string;
    putItem: DependencyEdge;
}

export type DependencyGraphDelete = {
    EphemeraId: string;
    deleteItem: DependencyEdge;
}

export type DependencyGraphAction = DependencyGraphPut | DependencyGraphDelete

export const isDependencyGraphPut = (action: DependencyGraphAction): action is DependencyGraphPut => ('putItem' in action)
export const isDependencyGraphDelete = (action: DependencyGraphAction): action is DependencyGraphDelete => ('deleteItem' in action)

export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
