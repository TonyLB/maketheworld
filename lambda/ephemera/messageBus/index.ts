import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'

import {
    MessageBus,
    isPublishMessage,
    isEphemeraUpdate,
    isDisconnectMessage,
    isConnectMessage,
    isWhoAmIMessage,
    isSyncRequest,
    isSyncResponse,
    isRegisterCharacterMessage,
    isFetchPlayerEphemera,
    isImportDefaults,
    isFetchImportDefaults,
    isPerception,
    isMoveCharacter,
    isDecacheAsset,
    isCacheAssetMessage,
    isPlayerUpdateMessage,
    isRoomUpdateMessage,
    isDescentUpdateMessage,
    isAncestryUpdateMessage,
    isDependencyCascadeMessage,
    isExecuteActionMessage,
    isMapSubscription,
    isMapUpdateMessage
} from "./baseClasses"

import publishMessage from '../publishMessage'
import ephemeraUpdate from '../ephemeraUpdate'
import disconnectMessage from '../disconnectMessage'
import connectMessage from '../connectMessage'
import whoAmIMessage from '../whoAmI'
import { syncRequest, syncResponse } from '../syncHandler'
import registerCharacter from '../registerCharacter'
import { fetchPlayerEphemera } from '../fetchEphemera'
import perceptionMessage from '../perception'
import moveCharacter from '../moveCharacter'
import decacheAssetMessage from '../decacheMessage'
import { cacheAssetMessage } from '../cacheAsset'
import playerUpdateMessage from '../playerUpdate'
import roomUpdateMessage from '../roomUpdate'
import dependentUpdateMessage from '../dependentMessages/dependentUpdate'
import dependencyCascadeMessage from '../dependentMessages/dependencyCascade'
import executeActionMessage from '../executeAction'
import mapSubscriptionMessage from '../mapSubscription'
import mapUpdateMessage from '../mapUpdate'

export const messageBus = new MessageBus()
messageBus.subscribe({
    tag: 'PublishMessage',
    priority: 15,
    filter: isPublishMessage,
    callback: publishMessage
})
messageBus.subscribe({
    tag: 'EphemeraUpdate',
    priority: 10,
    filter: isEphemeraUpdate,
    callback: ephemeraUpdate
})
messageBus.subscribe({
    tag: 'Disconnect',
    priority: 1,
    filter: isDisconnectMessage,
    callback: disconnectMessage
})
messageBus.subscribe({
    tag: 'Connect',
    priority: 1,
    filter: isConnectMessage,
    callback: connectMessage
})
messageBus.subscribe({
    tag: 'WhoAmI',
    priority: 1,
    filter: isWhoAmIMessage,
    callback: whoAmIMessage
})
messageBus.subscribe({
    tag: 'SyncResponse',
    priority: 2,
    filter: isSyncResponse,
    callback: syncResponse
})
messageBus.subscribe({
    tag: 'SyncRequest',
    priority: 3,
    filter: isSyncRequest,
    callback: syncRequest
})
messageBus.subscribe({
    tag: 'RegisterCharacter',
    priority: 1,
    filter: isRegisterCharacterMessage,
    callback: registerCharacter
})
messageBus.subscribe({
    tag: 'FetchPlayerEphemera',
    priority: 2,
    filter: isFetchPlayerEphemera,
    callback: fetchPlayerEphemera
})
messageBus.subscribe({
    tag: 'MapSubscription',
    priority: 2,
    filter: isMapSubscription,
    callback: mapSubscriptionMessage
})
messageBus.subscribe({
    tag: 'Perception',
    priority: 10,
    filter: isPerception,
    callback: perceptionMessage
})
messageBus.subscribe({
    tag: 'MoveCharacter',
    priority: 4,
    filter: isMoveCharacter,
    callback: moveCharacter
})
messageBus.subscribe({
    tag: 'DecacheAsset',
    priority: 1,
    filter: isDecacheAsset,
    callback: decacheAssetMessage
})
messageBus.subscribe({
    tag: 'CacheAsset',
    priority: 1,
    filter: isCacheAssetMessage,
    callback: cacheAssetMessage
})
messageBus.subscribe({
    tag: 'UpdatePlayer',
    priority: 5,
    filter: isPlayerUpdateMessage,
    callback: playerUpdateMessage
})
messageBus.subscribe({
    tag: 'RoomUpdate',
    priority: 6,
    filter: isRoomUpdateMessage,
    callback: roomUpdateMessage
})
messageBus.subscribe({
    tag: 'MapUpdate',
    priority: 15,
    filter: isMapUpdateMessage,
    callback: mapUpdateMessage
})
messageBus.subscribe({
    tag: 'AncestryUpdate',
    priority: 3,
    filter: isAncestryUpdateMessage,
    callback: dependentUpdateMessage('Ancestry')
})
messageBus.subscribe({
    tag: 'DescentUpdate',
    priority: 4,
    filter: isDescentUpdateMessage,
    callback: dependentUpdateMessage('Descent')
})

messageBus.subscribe({
    tag: 'DependencyCascade',
    priority: 5,
    filter: isDependencyCascadeMessage,
    callback: dependencyCascadeMessage
})

messageBus.subscribe({
    tag: 'ExecuteAction',
    priority: 7,
    filter: isExecuteActionMessage,
    callback: executeActionMessage
})

export default messageBus