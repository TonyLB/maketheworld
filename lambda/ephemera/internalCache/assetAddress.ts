import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { CacheConstructor } from './baseClasses'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/ts/readOnly'

export type AssetAddressItem = {
    EphemeraId: EphemeraAssetId;
    address: AssetWorkspaceAddress
}

export class CacheAssetAddressData {
    AssetAddressById: Record<EphemeraAssetId, AssetAddressItem> = {};
    clear() {
        this.AssetAddressById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<AssetAddressItem | undefined> {
        return this.AssetAddressById[assetId]
    }
    set(assetItem: AssetAddressItem): void {
        this.AssetAddressById[assetItem.EphemeraId] = assetItem
    }
}

export const CacheAssetAddress = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetAddress extends Base {
        AssetAddress: CacheAssetAddressData = new CacheAssetAddressData()

        override clear() {
            this.AssetAddress.clear()
            super.clear()
        }
    }
}

export default CacheAssetAddress
