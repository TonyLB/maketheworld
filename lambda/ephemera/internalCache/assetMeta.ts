import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetMetaItem = {
    EphemeraId: EphemeraAssetId;
    found: boolean;
}

export class CacheAssetMetaData {
    AssetMetaById: Record<EphemeraAssetId, AssetMetaItem> = {};
    clear() {
        this.AssetMetaById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetMetaItem, 'found'> | undefined> {
        if (!(this.AssetMetaById[assetId])) {
            const assetData = await ephemeraDB.getItem<AssetMetaItem>({
                    Key: {
                        EphemeraId: assetId,
                        DataCategory: 'Meta::Asset'
                    },
                    ProjectionFields: ['EphemeraId'],
                })
            if (assetData) {
                this.AssetMetaById[assetId] = {
                    ...assetData,
                    found: true
                }
            }
            else {
                this.AssetMetaById[assetId] = {
                    EphemeraId: assetId,
                    found: false
                }
            }
        }
        if (this.AssetMetaById[assetId]?.found) {
            const { found, ...rest } = this.AssetMetaById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetMetaItem, 'found'>): void {
        this.AssetMetaById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetMeta = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetMeta extends Base {
        AssetMeta: CacheAssetMetaData = new CacheAssetMetaData()

        override clear() {
            this.AssetMeta.clear()
            super.clear()
        }
    }
}

export default CacheAssetMeta
