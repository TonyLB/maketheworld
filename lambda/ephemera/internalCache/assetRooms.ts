import { EphemeraAssetId, EphemeraRoomId } from '@tonylb/mtw-interfaces/dist/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetRoomsItem = {
    EphemeraId: EphemeraAssetId;
    rooms: EphemeraRoomId[];
    found: boolean;
}

export class CacheAssetRoomsData {
    AssetRoomsById: Record<EphemeraAssetId, AssetRoomsItem> = {};
    clear() {
        this.AssetRoomsById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetRoomsItem, 'found'> | undefined> {
        if (!(this.AssetRoomsById[assetId])) {
            const assetRooms = await ephemeraDB.query<{ EphemeraId: EphemeraRoomId }[]>({
                IndexName: 'DataCategoryIndex',
                DataCategory: assetId,
                KeyConditionExpression: 'begins_with(EphemeraId, :roomPrefix)',
                ExpressionAttributeValues: {
                    ':roomPrefix': 'ROOM#'
                },        
            })
            if (assetRooms) {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: assetRooms.map(({ EphemeraId }) => (EphemeraId)),
                    found: true
                }
            }
            else {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: [],
                    found: false
                }
            }
        }
        if (this.AssetRoomsById[assetId]?.found) {
            const { found, ...rest } = this.AssetRoomsById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetRoomsItem, 'found'>): void {
        this.AssetRoomsById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetRooms = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetRooms extends Base {
        AssetRooms: CacheAssetRoomsData = new CacheAssetRoomsData()

        override clear() {
            this.AssetRooms.clear()
            super.clear()
        }
    }
}

export default CacheAssetRooms
