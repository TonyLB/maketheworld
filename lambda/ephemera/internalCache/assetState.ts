import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

type AssetStateAddress = {
    tag: 'Variable' | 'Computed';
    EphemeraId: string;
}

type AssetStateOutput<T extends Record<string, AssetStateAddress>> = {
    [key in keyof T]: any;
}

export class AssetStateData {
    _StatePromiseByEphemeraId: Record<string, Promise<any>> = {}
    _StateOverrides: Record<string, any> = {}
    
    clear() {
        this._StatePromiseByEphemeraId = {}
    }
    async get<T extends Record<string, AssetStateAddress>>(keys: T): Promise<AssetStateOutput<T>> {
        const itemsInNeedOfFetch = Object.values(keys)
            .filter(({ EphemeraId }) => (!(EphemeraId in this._StatePromiseByEphemeraId)))
        const batchGetPromise = ephemeraDB.batchGetItem<{ EphemeraId: string; value: any; }>({
            Items: itemsInNeedOfFetch.map(({ EphemeraId, tag }) => ({ EphemeraId, DataCategory: `Meta::${tag}` })),
            ProjectionFields: ['EphemeraId', '#value'],
            ExpressionAttributeNames: {
                '#value': 'value'
            }
        })
        Object.values(itemsInNeedOfFetch).forEach(({ EphemeraId }) => {
            this._StatePromiseByEphemeraId[EphemeraId] = batchGetPromise
                .then((items) => (items.find(({ EphemeraId: check }) => (check === EphemeraId))?.value))
                .catch((err) => {
                    if (typeof this._StateOverrides[EphemeraId] === 'undefined') {
                        throw err
                    }
                })
                .then((value) => ((typeof this._StateOverrides[EphemeraId] === 'undefined') ? value : this._StateOverrides[EphemeraId]))
        })
        return Object.assign({}, ...(await Promise.all(
            Object.entries(keys).map(async ([key, { EphemeraId }]) => ({ [key]: await this._StatePromiseByEphemeraId[EphemeraId] }))
        ))) as AssetStateOutput<T>
    }

    set(EphemeraId: string, value: any) {
        this._StateOverrides[EphemeraId] = value
    }

    invalidate(EphemeraId: string) {
        delete this._StateOverrides[EphemeraId]
        delete this._StatePromiseByEphemeraId[EphemeraId]
    }
}

export const AssetState = <GBase extends CacheConstructor>(Base: GBase) => {
    return class AssetState extends Base {
        AssetState: AssetStateData = new AssetStateData()

        override clear() {
            this.AssetState.clear()
            super.clear()
        }
    }
}

export default AssetState
