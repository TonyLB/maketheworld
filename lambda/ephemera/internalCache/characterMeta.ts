import { EphemeraCharacterId } from '@tonylb/mtw-interfaces/dist/baseClasses';
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { NormalCharacterPronouns } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { CacheConstructor } from './baseClasses'

export type CharacterMetaItem = {
    EphemeraId: EphemeraCharacterId;
    Name: string;
    RoomId: string;
    Color?: string;
    fileURL?: string;
    HomeId: string;
    assets: string[];
    Pronouns: NormalCharacterPronouns;
}

export class CacheCharacterMetaData {
    CharacterMetaById: Record<EphemeraCharacterId, CharacterMetaItem> = {};
    clear() {
        this.CharacterMetaById = {}
    }
    async get(characterId: EphemeraCharacterId): Promise<CharacterMetaItem> {
        if (!(this.CharacterMetaById[characterId])) {
            const characterData = await ephemeraDB.getItem<Omit<CharacterMetaItem, 'EphemeraId'>>({
                    EphemeraId: characterId,
                    DataCategory: 'Meta::Character',
                    ProjectionFields: ['#name', 'RoomId', 'Color', 'fileURL', 'HomeId', 'assets', 'Pronouns'],
                    ExpressionAttributeNames: {
                        '#name': 'Name'
                    }
                }) || { Name: '', RoomId: '', Color: 'grey', fileURL: '', HomeId: 'VORTEX', assets: [], Pronouns: { subject: 'they', object: 'them', possessive: 'their', adjective: 'theirs', reflexive: 'themself' } }
            this.CharacterMetaById[characterId] = {
                ...characterData,
                EphemeraId: characterId
            }
        }
        return this.CharacterMetaById[characterId]
    }
}

export const CacheCharacterMeta = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheCharacterMeta extends Base {
        CharacterMeta: CacheCharacterMetaData = new CacheCharacterMetaData()

        override clear() {
            this.CharacterMeta.clear()
            super.clear()
        }
    }
}

export default CacheCharacterMeta
