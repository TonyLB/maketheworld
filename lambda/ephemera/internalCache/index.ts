import { CacheConstructor, CacheBase } from './baseClasses'

import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'
import { delayPromise } from '@tonylb/mtw-utilities/dist/dynamoDB/delayPromise'
import CacheRoomCharacterLists from './roomCharacterLists';
import CacheCharacterMeta from './characterMeta';
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB';
import CacheCharacterConnections from './characterConnections';
import AssetState from './assetState';
import DependencyGraph from './dependencyGraph';
import ComponentMeta from './componentMeta';
import ComponentRender from './componentRender';

type CacheGlobalKeys = 'ConnectionId' | 'RequestId' | 'player' | 'assets' | 'connections'
class CacheGlobalData {
    ConnectionId?: string;
    RequestId?: string;
    player?: string;
    assets?: string[];
    connections?: string[];
    get(key: 'ConnectionId' | 'RequestId' | 'player'): Promise<string | undefined>
    get(key: 'assets' | 'connections'): Promise<string[] | undefined>
    get(key: CacheGlobalKeys): Promise<string | string[] | undefined>
    async get(key: CacheGlobalKeys) {
        switch(key) {
            case 'player':
                if (!this.player && this.ConnectionId) {
                    //
                    // TODO: Replace repeated attempts with exponential backoff by
                    // refactoring connectionDB.getItem to allow a consistent argument
                    // that can actviate strongly-consistent reads
                    //
                    let attempts = 0
                    let exponentialBackoff = 50
                    while(attempts < 5) {
                        const { player = '' } = await connectionDB.getItem<{ player: string }>({
                            ConnectionId: `CONNECTION#${this.ConnectionId}`,
                            DataCategory: 'Meta::Connection',
                            ProjectionFields: ['player']
                        }) || {}
                        if (player) {
                            this.player = player
                            return player
                        }
                        attempts += 1
                        await delayPromise(exponentialBackoff)
                        exponentialBackoff = exponentialBackoff * 2
                    }
                    console.log(`Exponential backoff on player caching failed after five attempts (${this.ConnectionId})`)
                }
                return this.player
            case 'assets':
                if (typeof this.assets === 'undefined') {
                    const { assets = [] } = (await ephemeraDB.getItem<{ assets: string[] }>({
                        EphemeraId: 'Global',
                        DataCategory: 'Assets',
                        ProjectionFields: ['assets']
                    })) || {}
                    this.assets = assets
                }
                return this.assets
            case 'connections':
                if (typeof this.connections === 'undefined') {
                    const { connections = {} } = (await connectionDB.getItem<{ connections: Record<string, string> }>({
                        ConnectionId: 'Global',
                        DataCategory: 'Connections',
                        ProjectionFields: ['connections']
                    })) || {}
                    this.connections = Object.keys(connections)
                }
                return this.connections
            default:
                return this[key]
        }
    }

    clear() {
        this.ConnectionId = undefined
        this.RequestId = undefined
        this.player = undefined
        this.assets = undefined
    }

    set(props: { key: 'ConnectionId' | 'RequestId', value: string; }): void {
        this[props.key] = props.value
    }
}

export const CacheGlobal = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Global: CacheGlobalData = new CacheGlobalData()

        override clear() {
            this.Global.clear()
            super.clear()
        }
    }
}

const InternalCache = ComponentRender(ComponentMeta(AssetState(DependencyGraph(CacheCharacterConnections(CacheCharacterMeta(CacheRoomCharacterLists(CacheGlobal(CacheBase))))))))
export const internalCache = new InternalCache()
export default internalCache
