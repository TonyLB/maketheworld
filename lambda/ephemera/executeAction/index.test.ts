jest.mock('../messageBus')
import messageBus from '../messageBus'

jest.mock('@tonylb/mtw-utilities/dist/dynamoDB')
import { ephemeraDB, multiTableTransactWrite, exponentialBackoffWrapper } from "@tonylb/mtw-utilities/dist/dynamoDB"

jest.mock('../internalCache')
import internalCache from '../internalCache'

import executeActionMessage from '.'
import { CharacterMetaItem } from '../internalCache/characterMeta'

const ephemeraDBMock = ephemeraDB as jest.Mocked<typeof ephemeraDB>
const messageBusMock = messageBus as jest.Mocked<typeof messageBus>
const internalCacheMock = jest.mocked(internalCache, true)
const transactMock = multiTableTransactWrite as jest.Mock
const exponentialBackoffMock = exponentialBackoffWrapper as jest.Mock

describe('ExecuteActionMessage', () => {
    const testCharacterMeta: CharacterMetaItem = {
        Name: 'Tess',
        Pronouns: { subject: 'she', object: 'her', possessive: 'her', adjective: 'hers', reflexive: 'herself' },
        Color: 'purple',
        EphemeraId: 'CHARACTER#Somebody',
        RoomId: 'VORTEX',
        HomeId: 'VORTEX'
    }
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        internalCache.clear()
        exponentialBackoffMock.mockImplementation(async (func) => { await func() })
    })

    it('should execute an action and cascade', async () => {
        ephemeraDBMock.getItem.mockResolvedValue({
            src: 'a = a + b',
            rootAsset: 'base'
        })
        ephemeraDBMock.query.mockResolvedValue([])
        internalCacheMock.AssetMap.get.mockResolvedValue({
            a: 'VARIABLE#VariableOne',
            b: 'VARIABLE#VariableTwo'
        })
        internalCacheMock.AssetState.get.mockResolvedValue({
            a: 1,
            b: 2
        })
        internalCacheMock.CharacterMeta.get.mockResolvedValue(testCharacterMeta)
        await executeActionMessage({
            payloads: [
                { type: 'ExecuteAction', actionId: 'ACTION#TestOne', characterId: 'CHARACTER#Somebody' },
                { type: 'ExecuteAction', actionId: 'ACTION#TestTwo', characterId: 'CHARACTER#SomebodyElse' },
            ],
            messageBus: messageBusMock
        })
        expect(transactMock).toHaveBeenCalledTimes(1)
        expect(transactMock.mock.calls[0][0]).toMatchSnapshot()
        expect(messageBusMock.send).toHaveBeenCalledTimes(2)
        expect(messageBusMock.send.mock.calls.map(([item]) => (item))).toMatchSnapshot()

    })

})