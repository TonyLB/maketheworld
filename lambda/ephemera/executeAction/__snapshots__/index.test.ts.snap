// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteActionMessage should execute an action and cascade 1`] = `
Array [
  Object {
    "ConditionCheck": Object {
      "ConditionExpression": "#value = :value",
      "ExpressionAttributeValues": Object {
        ":value": 2,
      },
      "Key": Object {
        "DataCategory": "Meta::Variable",
        "EphemeraId": "VARIABLE#VariableTwo",
      },
      "ProjectionFields": Array [
        "value",
      ],
    },
  },
]
`;

exports[`ExecuteActionMessage should execute an action and cascade 2`] = `
Object {
  "actionId": "ACTION#TestTwo",
  "characterId": "CHARACTER#SomebodyElse",
  "type": "ExecuteAction",
}
`;
